sweepbatcher/sweep_batch.go

```diff
@@ -1365,15 +1398,96 @@ func constructUnsignedTx(sweeps []sweep, address btcutil.Address,
 		feeForWeight++
 	}
 
+	// Add the batch transaction output, which excludes the fees paid to
+	// miners. Reduce the amount by the sum of change outputs, if any.
+	var sumChange int64
+	for _, change := range changeOutputs {
+		sumChange += change.Value
+	}
+
+	// Ensure that the batch amount is greater than the sum of change.
+	if batchAmt <= btcutil.Amount(sumChange) {
+		return nil, 0, 0, 0, fmt.Errorf("batch amount %v is <= the "+
+			"sum of change outputs %v", batchAmt,
+			btcutil.Amount(sumChange))
+	}
+
 	// Clamp the calculated fee to the max allowed fee amount for the batch.
-	fee := clampBatchFee(feeForWeight, batchAmt)
+	fee, err := clampBatchFee(
+		feeForWeight, batchAmt-btcutil.Amount(sumChange),
+		minRelayFeeRate, weight,
+	)
+	if err != nil {
+		return nil, 0, 0, 0, fmt.Errorf("failed to clamp batch "+
+			"fee: %w", err)
+	}
 
-	// Add the batch transaction output, which excludes the fees paid to
-	// miners.
+	// Ensure that batch amount is equal or exceeds the sum of change
+	// outputs and the fee, and that it is also greater than dust limit
+	// for the main output.
+	dustLimit := utils.DustLimitForPkScript(batchPkScript)
+	if fee+btcutil.Amount(sumChange)+dustLimit > batchAmt {
+		return nil, 0, 0, 0, fmt.Errorf("batch amount %v is < the "+
+			"sum of change outputs %v plus fee %v and dust "+
+			"limit %v", batchAmt, btcutil.Amount(sumChange),
+			fee, dustLimit)
+	}
+
+	// Add the main output first.
 	batchTx.AddTxOut(&wire.TxOut{
 		PkScript: batchPkScript,
-		Value:    int64(batchAmt - fee),
+		Value:    int64(batchAmt-fee) - sumChange,
 	})
+	// Then add change outputs. Sort the keys first to make tests
+	// deterministic.
+	keys := make([]string, 0, len(changeOutputs))
+	for k := range changeOutputs {
+		keys = append(keys, k)
+	}
+
+	// Sort the keys
+	sort.Strings(keys)
```

Comment:
Let's use [BIP-0069](https://github.com/bitcoin/bips/blob/master/bip-0069.mediawiki) for outputs sorting (change only, so main output is 0).
We have it implemented for inputs in `utils/get_prevout_info.go`